File: Python/algorithmic_heights/09_bfs.py
# Double-Degree Array
import sys

from collections import defaultdict, deque


class DirectedGraph:
    def __init__(self, nr_nodes, edge_list):
        self._nr_nodes = nr_nodes
        self._adjacency_list = self._build_adjacency_list(edge_list)

    @staticmethod
    def _build_adjacency_list(edge_list):
        adjacency_list = defaultdict(list)
        for a, b in edge_list:
            adjacency_list[a].append(b)
        return dict(adjacency_list)

    def breadth_first_search(self, start_node):
        distances = [-1] * self._nr_nodes
        distances[start_node-1] = 0
        queue = deque([start_node])
        while queue:
            node = queue.popleft()
            neighbours = self._adjacency_list.get(node, [])
            for neighbour in neighbours:
                if distances[neighbour-1] == -1:
                    queue.append(neighbour)
                    distances[neighbour-1] = distances[node-1] + 1
        return distances


def convert_to_intlist(line):
    return tuple(int(item) for item in line.split())


def read_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        a, b = convert_to_intlist(next(reader))
        edges.append((a, b))
    return edges


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_edges(reader, nr_edges)
    graph = DirectedGraph(nr_nodes, edge_list)
    result = graph.breadth_first_search(start_node=1)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/09_bfs.hs
import Control.Monad (replicateM)
import Data.IntMap (IntMap, (!))
import Data.Maybe (fromMaybe)
import Data.Sequence ((|>), ViewL ((:< ), EmptyL))
import qualified Data.IntMap as Map
import qualified Data.Sequence as S

type AdjacencyList = IntMap [Int]
data Edge = Edge Int Int
data DirectedGraph = DirectedGraph { nrNodes :: Int, adjacencyList :: AdjacencyList }

convertToIntList :: String -> [Int]
convertToIntList = map read . words


readInputLine :: String -> (Int, Int)
readInputLine line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


readEdges :: Int -> IO [Edge]
readEdges nrEdges = do
    edges <- replicateM nrEdges getLine
    return $ map (\line -> let (a, b) = readInputLine line in Edge a b) edges


createGraph :: Int -> [Edge] -> DirectedGraph
createGraph nrNodes edges =
    let addEdge adjacencyList (Edge a b) =
            let update Nothing = Just [b]
                update (Just ns) = Just (b : ns)
            in Map.alter update a adjacencyList
        adjacencyList = foldl addEdge Map.empty edges
    in DirectedGraph { nrNodes = nrNodes, adjacencyList = adjacencyList }


getNeighbors :: DirectedGraph -> Int -> [Int]
getNeighbors (DirectedGraph _ adjacencyList) node =
    fromMaybe [] $ Map.lookup node adjacencyList


breadthFirstSearch :: DirectedGraph -> Int -> [Int]
breadthFirstSearch graph @ (DirectedGraph nrNodes adjacencyList) startNode =
    let go queue distances =
            case S.viewl queue of
                EmptyL -> map (\node -> Map.findWithDefault (-1) node distances) [1..nrNodes]
                (node :< rest) ->
                    let nodeDistance = distances ! node
                        update [] q dists = (q, dists)
                        update (neighbor : ns) q dists =
                            if Map.member neighbor dists then update ns q dists
                            else update ns (q |> neighbor) (Map.insert neighbor (nodeDistance + 1) dists)
                        neighbors = getNeighbors graph node
                        (queue', distances') = update neighbors rest distances
                    in go queue' distances'
    in go (S.singleton startNode) (Map.singleton startNode 0)


main :: IO ()
main = do
    (nrNodes, nrEdges) <- readInputLine <$> getLine
    edges <- readEdges nrEdges
    let graph = createGraph nrNodes edges
    let result = breadthFirstSearch graph 1
    putStrLn $ unwords $ map show result
